### ISSUE TITLE
Android 8 Push notification not working

### ISSUE THREAD

**Issue by shaktisinghtanwar** [myflashlab/Firebase-ANE#264]
## Environment ##
* Android Version: 8.1
* iOS Version: 12
* AIR Version: 4.7
AIRSDK 31

## Problem Description ##
I have implemented firebasemessaging and firebasecore ANE in my adobe flex app using AS3. after implementation i got the push notification in android 6 and 7 but NOT in ANDROID 8. can you please suggest me what do to get push notification in android 8. 
```xml
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<application xmlns="http://ns.adobe.com/air/application/30.0">


	<android>
        <colorDepth>16bit</colorDepth>
        <containsVideo>true</containsVideo>
        <manifestAdditions><![CDATA[
			<manifest android:installLocation="auto">
				<uses-sdk android:targetSdkVersion="26"/>
				<uses-sdk android:minSdkVersion="9"/>
				<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
				<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
				<uses-permission android:name="android.permission.INTERNET"/>
				<uses-permission android:name="android.permission.READ_PHONE_STATE" />
				<uses-permission android:name="android.permission.GET_ACCOUNTS"/>
				<uses-permission android:name="android.permission.GET_TASKS"/>
				<uses-permission android:name="android.permission.VIBRATE"/>
				<uses-permission android:name="android.permission.WAKE_LOCK"/>


				<permission android:name="air.com.xxxxx.permission.C2D_MESSAGE" android:protectionLevel="signature" />
				<uses-permission android:name="air.com.xxxxx.permission.C2D_MESSAGE" />
				<uses-permission android:name="com.google.android.c2dm.permission.RECEIVE" />
				<application>
					<service android:name=".NotificationListener"
         			android:permission="android.permission.BIND_NOTIFICATION_LISTENER_SERVICE">
				     <intent-filter>
				         <action android:name="android.service.notification.NotificationListenerService" />
				     </intent-filter>
					 </service>
	
				<receiver
                    android:name="com.google.firebase.iid.FirebaseInstanceIdReceiver"
                    android:exported="true"
                    android:permission="com.google.android.c2dm.permission.SEND">
                    <intent-filter>
                        <action android:name="com.google.android.c2dm.intent.RECEIVE"/>
                        <category android:name="air.com.xxxxx.firebaseCore"/>
                    </intent-filter>
                </receiver>
                <service
                    android:name="com.google.firebase.iid.FirebaseInstanceIdService"
                    android:exported="true">
                    <intent-filter android:priority="-500">
                        <action android:name="com.google.firebase.INSTANCE_ID_EVENT"/>
                    </intent-filter>
                </service>
					
					<activity android:launchMode="standard">
						<intent-filter> 
						 <action android:name="android.intent.action.MAIN"/> 
						 <category android:name="android.intent.category.LAUNCHER"/> 
						</intent-filter> 
						<intent-filter>
							<action android:name="android.intent.action.VIEW" />
							<category android:name="android.intent.category.DEFAULT" />
							<category android:name="android.intent.category.BROWSABLE" />
							 <data android:scheme="air.com.xxxxx.firebaseCore" />
						</intent-filter>
					</activity>
				
                    <service android:name="com.myflashlab.firebase.fcm.MyFirebaseMessagingService">
					<intent-filter>
						<action android:name="com.google.firebase.MESSAGING_EVENT"/>
					</intent-filter>
				</service>
				
				<activity 
					android:name="com.myflashlab.firebase.fcm.MyFirebaseMessagingIntent" 
					android:exported="false"
					android:noHistory="true"
					android:clearTaskOnLaunch="true"
					android:hardwareAccelerated="false"
					android:stateNotNeeded="true">
					<intent-filter>
						<action android:name="MY_INTENT" />
						<category android:name="android.intent.category.DEFAULT" />
					</intent-filter>
				</activity>
					
					<!-- this meta-data tag is required for GAnalytics -->					
					<meta-data android:name="com.google.android.gms.version" android:value="4452000"/>
					

	 			<activity android:name="com.facebook.LoginActivity" android:theme="@android:style/Theme.Translucent.NoTitleBar" android:label="Login" />
				  <!-- replace 000000000000000 with your numeric Facebook Application ID -->
				  <provider android:authorities="com.facebook.app.NativeAppCallContentProviderxxxxxxxxx" android:name="com.facebook.NativeAppCallContentProvider" />
				  <!-- replace 000000000000000 with your numeric Facebook Application ID, and com.yourcompany.example with your apks application ID -->
				  <meta-data android:name="air.com.yourcompany.example.ApplicationId" android:value="xxxxxxxxxxxxxxxxxxxxxx"/>
				</application>	
			</manifest>
			
		]]></manifestAdditions>
    </android>

     <iPhone>
      <Entitlements> 
	      <![CDATA[ 
	         <key>aps-environment</key> 
	         <string>production</string>  <!-- production development -->
	      ]]> 
	   </Entitlements> 
        <InfoAdditions><![CDATA[
        	<key>NSAppTransportSecurity</key>
			<dict>
			    <key>NSAllowsArbitraryLoads</key>
			    <true/>
			</dict>
        	<key>UIApplicationExitsOnSuspend</key> 
        		<false/>
	        <key>UIViewControllerBasedStatusBarAppearance</key> 
	 			<false/> 
	        <key>UIStatusBarStyle</key> 
	        	<string>UIStatusBarStyleDefault</string>
	   		<key>UIDeviceFamily</key>
	   			<array>
	    			<string>1</string>
	    			<string>2</string>
	   			</array>
	   		<key>FacebookAppID</key>
				<string>xxxxxxxxxxxxxx</string>
			<key>FacebookDisplayName</key>
				<string>The xxxxx</string>
		 	<key>MinimumOSVersion</key> 
	    		<string>6.0</string>
	   		<key>CFBundleURLTypes</key>
	    		<array>
	     			<dict>
	     				<key>CFBundleURLName</key>
						<string>com.xxxxx</string>
	      				<key>CFBundleURLSchemes</key>
	      				<array>
	       					<string>xxxxxxxxxxxxxxx</string>
	      				</array>
	     			</dict>
	    		</array>
	    		
  			]]></InfoAdditions>
        <requestedDisplayResolution excludeDevices="iPad1,1 iPad2,1 iPad2,2 iPad2,3 iPad2,4 iPad2,5 iPad3,1 iPad3,2 iPad3,3 iPad3,4 iPad4,1 iPad4,2 iPad4,3 iPad4,4 iPad4,5 iPad4,6">high</requestedDisplayResolution>
    </iPhone>
    <supportedLanguages>en fr es</supportedLanguages>
    
	<extensions>
	<extensionID>com.myflashlab.air.extensions.firebase.core</extensionID>
	
		<extensionID>com.milkmangames.extensions.EasyPush</extensionID>
	    <extensionID>com.milkmangames.extensions.GoViral</extensionID>
	    <extensionID>com.myflashlab.air.extensions.dependency.overrideAir</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.androidSupport.arch</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.androidSupport.core</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.androidSupport.v4</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.firebase.analytics</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.firebase.analytics.impl</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.firebase.common</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.firebase.iid</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.firebase.measurement.connector</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.firebase.measurement.connector.impl</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.ads</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.base</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.basement</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.measurementBase</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.stats</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.tasks</extensionID>
	    <extensionID>com.azer.IOSApplicationSettings</extensionID>
	    <extensionID>com.doitflash.air.extensions.videoPlayer</extensionID>
    	<extensionID>com.freshplanet.AirAlert</extensionID>
    <extensionID>com.freshplanet.AirPushNotification</extensionID>
    <extensionID>com.mesmotronic.ane.androidid</extensionID>
    <extensionID>com.milkmangames.extensions.GAnalytics</extensionID>
        <extensionID>com.milkmangames.extensions.GoogleServices</extensionID>
    <extensionID>com.myflashlab.air.extensions.firebase.fcm</extensionID>
    </extensions>
</application>
```

```actionscript
// ActionScript file
package com
// ActionScript file
package com
{

import com.mesmotronic.ane.AndroidID;
import com.milkmangames.nativeextensions.events.PNEvent;
import com.myflashlab.air.extensions.dependency.OverrideAir;
import com.myflashlab.air.extensions.firebase.core.Firebase;
import com.myflashlab.air.extensions.firebase.core.FirebaseConfig;
import com.myflashlab.air.extensions.firebase.fcm.FCM;
import com.myflashlab.air.extensions.firebase.fcm.FcmEvents;
import com.doitflash.starling.utils.list.List;
import com.doitflash.text.modules.MySprite;
import com.doitflash.mobileProject.commonCpuSrc.DeviceInfo;
import flash.text.AntiAliasType;
import com.luaye.console.C;


import flash.desktop.NativeApplication;
import flash.desktop.SystemIdleMode;
import flash.display.Sprite;
import flash.display.StageAlign;
import flash.display.StageScaleMode;
import flash.events.Event;
import flash.events.MouseEvent;
import flash.media.StageWebView;
import flash.system.Capabilities;
import flash.text.TextField;
import flash.text.TextFieldAutoSize;
import flash.text.TextFormat;
import flash.text.TextFormatAlign;

import views.AppWebViewScreenView;
import views.CookingMethodDetailScreenView;
import views.CutDetailScreenView;
import views.RecipesDetailScreenView;
import views.cmsPagesDetailView;
import views.cmsPagesListView;

[SWF(frameRate="40", backgroundColor="#000000")]

public class AndroidNotification extends Sprite
{
	public var DeeplinkingObj:Object = new Object(); 
	public static const GCM_SENDER_ID:String = "xxxxxxxxxxxxxx";
	private var _navigator:Object;
	private var _list:List;
	
	private const BTN_WIDTH:Number = 150;
	private const BTN_HEIGHT:Number = 60;
	private const BTN_SPACE:Number = 2;
	private var _txt:TextField;
	private var _body:Sprite;
	private var _numRows:int = 1;
	
	public function AndroidNotification()
	{
		NativeApplication.nativeApplication.addEventListener(Event.ACTIVATE, handleActivate);
		NativeApplication.nativeApplication.addEventListener(Event.DEACTIVATE, handleDeactivate);
		
		init();

	}

	private function handleActivate(e:Event):void
	{
		NativeApplication.nativeApplication.systemIdleMode = SystemIdleMode.KEEP_AWAKE;
	}
	
	private function handleDeactivate(e:Event):void
	{
		NativeApplication.nativeApplication.systemIdleMode = SystemIdleMode.NORMAL;
	}

	private function myDebuggerDelegate($ane:String, $class:String, $msg:String):void
	{
		trace($ane + "(" + $class + ")" + " " + $msg);
	}

		 private function init():void
		 {
			 // remove this line in production build or pass null as the delegate
			 OverrideAir.enableDebugger(myDebuggerDelegate);
			 
			 var isConfigFound:Boolean = Firebase.init();
			 
			 if (isConfigFound)
			 {
				 var config:FirebaseConfig = Firebase.getConfig();
								 
				 // FCM needs Google Services so, before using FCM, we need to check that first.
				 // https://firebase.google.com/docs/cloud-messaging/android/client#sample-play
				 if (Firebase.os == Firebase.ANDROID) Firebase.checkGoogleAvailability(onCheckResult);
				 else onCheckResult(Firebase.SUCCESS);
				 
			 }

			 else
			 {
				 trace("Config file is not found!");
			 }
		 }
	 
	
		 private function onCheckResult($result:int):void
		 {
			 switch($result)
			 {
				 case Firebase.SUCCESS:
					 
					 trace("checkGoogleAvailability result = SUCCESS");
					 
					 // now you can use FCM
					 initFCM();

					 break;
				 case Firebase.SERVICE_MISSING:
					 
					 trace("checkGoogleAvailability result = SERVICE_MISSING");
					 
					 break;
				 case Firebase.SERVICE_VERSION_UPDATE_REQUIRED:
					 
					 trace("checkGoogleAvailability result = SERVICE_VERSION_UPDATE_REQUIRED");
					 
					 break;
				 case Firebase.SERVICE_UPDATING:
					 
					 trace("checkGoogleAvailability result = SERVICE_UPDATING");
					 
					 break;
				 case Firebase.SERVICE_DISABLED:
					 
					 trace("checkGoogleAvailability result = SERVICE_DISABLED");
					 
					 break;
				 case Firebase.SERVICE_INVALID:
					 
					 trace("checkGoogleAvailability result = SERVICE_INVALID");
					 
					 break;
			 }
		 }

		 private function initFCM():void
		 {
			 FCM.init(); 
			 FCM.listener.addEventListener(FcmEvents.TOKEN_REFRESH, onTokenRefresh);
			 FCM.listener.addEventListener(FcmEvents.MESSAGE, onMessage);
			 
			 var btn1:MySprite = createBtn("getToken");
			 btn1.addEventListener(MouseEvent.CLICK, getToken);
			 _list.add(btn1);
 
			 function getToken(e:MouseEvent):void
			 {
				 FCM.getInstanceId(onTokenReceived);
			 }
			 
			 function onTokenReceived($token:String, $error:String):void
			 {
				 if($error)
				 {
					 trace("onTokenReceived error: " + $error);
					
				 }
				 
				 if($token)
				 {
					 trace("token: " + $token);
					
				 }
			 }
	 
		 }
		 private function onTokenRefresh(e:FcmEvents):void
		 {
			 trace("onTokenRefresh = " + e.token);
		 }

		 private function onMessage(e:FcmEvents):void
		 {
			 trace(e.msg);
			 var payload:Object = FCM.parsePayloadFromString(e.msg);
			 
			 if (payload)
			 {
				 for (var name:String in payload)
				 {
					 trace(name + " = " + payload[name]);
				 }
			 }
		 }
		
		 
		 private function createBtn($str:String):MySprite
		 {
			 var sp:MySprite = new MySprite();
			 sp.addEventListener(MouseEvent.MOUSE_OVER,  onOver);
			 sp.addEventListener(MouseEvent.MOUSE_OUT,  onOut);
			 sp.addEventListener(MouseEvent.CLICK,  onOut);
			 sp.bgAlpha = 1;
			 sp.bgColor = 0xDFE4FF;
			 sp.drawBg();
			 sp.width = BTN_WIDTH * DeviceInfo.dpiScaleMultiplier;
			 sp.height = BTN_HEIGHT * DeviceInfo.dpiScaleMultiplier;
			 
			 function onOver(e:MouseEvent):void
			 {
				 sp.bgAlpha = 1;
				 sp.bgColor = 0xFFDB48;
				 sp.drawBg();
			 }
			 
			 function onOut(e:MouseEvent):void
			 {
				 sp.bgAlpha = 1;
				 sp.bgColor = 0xDFE4FF;
				 sp.drawBg();
			 }
			 
			 var format:TextFormat = new TextFormat("Arimo", 16, 0x666666, null, null, null, null, null, TextFormatAlign.CENTER);
			 
			 var txt:TextField = new TextField();
			 txt.autoSize = TextFieldAutoSize.LEFT;
			 txt.antiAliasType = AntiAliasType.ADVANCED;
			 txt.mouseEnabled = false;
			 txt.multiline = true;
			 txt.wordWrap = true;
			 txt.scaleX = txt.scaleY = DeviceInfo.dpiScaleMultiplier;
			 txt.width = sp.width * (1 / DeviceInfo.dpiScaleMultiplier);
			 txt.defaultTextFormat = format;
			 txt.text = $str;
			 
			 txt.y = sp.height - txt.height >> 1;
			 sp.addChild(txt);
			 
			 return sp;
		 }
		 

}



}
```
## Steps to Reproduce ##
please suggest

## Known Workarounds ##)

**Comment by myflashlab** [comment]
We confirm that the FCM ANE works perfectly fine on Android 8+ as we do our tests mostly on the latest OS. However, there are a few things to notice:

1. You are using another pushNotification service ANE, that might create conflicts in unexpected ways because notifications API rules has been changed with Android 8 and if other libraries are not updated with the latest changes, they would probably make troubles. So the first thing I would do is to remove other unrelated ANEs and see how it goes.
2. The second thing I would try is to watch out for Logcat traces and check if I can see any logs realated to FCM when I send one from the server. Maybe there's an error or something when the message arrives? If so, post that error message here so we can investigate the problem. But if you don't see anything happening in the logcat traces, well, that means the FCM is not even reaching the device!

Try the above two approaches and let us know how you go.

**Comment by shaktisinghtanwar** [comment]
above is my code after removing other ANE's, still not getting any push notification. the function init() have if and else condition. while debugging it goes into the if condition and do not return any thing. so i am not getting the device token.
 this is the log

**Comment by myflashlab** [comment]
Your logs are saying that the ```onCheckResult``` is not called after ```Firebase.checkGoogleAvailability(onCheckResult);``` happens.

First remove the unnecessary trace: ```trace(Firebase.checkGoogleAvailability(onCheckResult));```
and then open up Logcat window in AndroidStudio and watch out what happens when that method is called. 

This is the first time that someone is reporting such a thing :thinking: we are interested to see what might be going wrong there.

Also tell me if you can test on another device and see if you are experiencing the same issue?

**Comment by shaktisinghtanwar** [comment]
After removing trace i got this. For the proper implementation i have to call fcm.init() function. And according to your sample code fcm.init runs when else condition become true


com.myflashlab.air.extensions.firebase.core(AirCommand) command:initFirebase
com.myflashlab.air.extensions.firebase.core(AirCommand) initFirebase > passed the initial initialization
com.myflashlab.air.extensions.firebase.core(AirCommand) initFirebase > initialization completed
com.myflashlab.air.extensions.firebase.core(AirCommand) command:initFirebase
com.myflashlab.air.extensions.firebase.core(AirCommand) initFirebase > passed the initial initialization
com.myflashlab.air.extensions.firebase.core(AirCommand) initFirebase > initialization completed

**Comment by myflashlab** [comment]
the ```initFirebase > passed the initial initialization``` is happening twice! Are you trying to initialize the ANE twice? please re-check your code and make sure it's not happening.

And besides, watching the AndroidStudio logcat traces could be very helpful to understand what could be going wrong. If you don't know how to open the logcat window, here it is:

1. install AndroidStudio
2. open a dummy new project
3. open the logcat window while running the AIR app in your device and the device is connected to your comouter.

**Comment by myflashlab** [comment]
Great, also double check your code with this sample one: https://github.com/myflashlab/Firebase-ANE/blob/master/AIR/src/MainFcm.as#L160

**Comment by shaktisinghtanwar** [comment]
this is the logcat file from android studio. i have rechecked the code from your given sample and still if condition becomes true
[logcat.txt](https://github.com/myflashlab/Firebase-ANE/files/2614490/logcat.txt)

**Comment by myflashlab** [comment]
Ok, I found:
```
java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.tasks.OnSuccessListener" on path: DexPathList[[zip file "/data/app/air.com.roundup.debug-SlkUkZjDjxT4psBktxcauA==/base.apk"],nativeLibraryDirectories=[/data/app/air.com.roundup.debug-SlkUkZjDjxT4psBktxcauA==/lib/arm, /data/app/air.com.roundup.debug-SlkUkZjDjxT4psBktxcauA==/base.apk!/lib/armeabi-v7a, /system/lib, /system/vendor/lib]]
```

The class ```com.google.android.gms.tasks.OnSuccessListener``` is available in the following ANE:
```xml
<extensionID>com.myflashlab.air.extensions.dependency.googlePlayServices.tasks</extensionID>
```

in your first log I can see that you have implemented this ANE in your project. So it's really a wonder why it says it's not found.

Are you sure the very same .apk file works on other Android versions? If that class is missing, all the other devices should fail too!

------------
At the beginning of the log, I also noticed: ```[GetToken] GetToken failed with status code: BadAuthentication``` This is again an interesting one! I [searched google](https://www.google.com/search?q=%5BGetToken%5D+GetToken+failed+with+status+code%3A+BadAuthentication&ie=utf-8&oe=utf-8&client=firefox-b-ab) with it and I can say whatever it is, it is not an ANE related issue.

yet, it seems like it's not even related to your app. it's just something else going wrong on your device prior to your app being started. So, I guess this can be ignore. just mentioned it here for your notice.

------------

1. Are you experiencing a crash in your app also?
2. Are you using all the latest versions of all ANEs? please confirm that.

So, if you are confirming above, let's get down to this by more deep tests:
1. Create a new project with the same package name you have for your app.
2. In the new project init firebase core and fcm ONLY, just like the sample project we have on our github
3. send us the compiled .apk file of this project so we can run it here locally
4. Also send us this demo project (**do NOT post the ANEs**)

**Comment by shaktisinghtanwar** [comment]
Ok, i will send you a dummy project

**Comment by myflashlab** [comment]
Ok, I guessed right! I just opened up your .apk file and noticed that you have NOT implemented the dependency ANEs! Just add them and the problem ```java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.tasks.OnSuccessListener"``` will be fixed.

<img width="1280" alt="screen shot 2018-11-27 at 2 51 31 pm" src="https://user-images.githubusercontent.com/1692948/49077228-3549be00-f254-11e8-9884-7d0fbf173705.png">

If you have other questions, feel free to open a new ticket. I'm closing this one because the reason is found and solved.
